netsync/manager_test.go

```diff
@@ -0,0 +1,776 @@
+			OnReject: func(p *peer.Peer, msg *wire.MsgReject) {
+				remoteMessages.rejectChan <- msg
+			},
+		},
+		UserAgentName:    "btcdtest",
+		UserAgentVersion: "1.0",
+		ChainParams:      &chainParams,
+		Services:         wire.SFNodeNetwork,
+	}
+
+	localPeerCfg := peer.Config{
+		UserAgentName:    "btcdtest",
+		UserAgentVersion: "1.0",
+		ChainParams:      &chainParams,
+		Services:         wire.SFNodeNetwork,
+	}
+
+	_, localNode, err :=
+		MakeConnectedPeers(remotePeerCfg, localPeerCfg, 0)
+	if err != nil {
+		t.Fatal(err)
+	}
+	syncMgr.NewPeer(localNode, nil)
+
+	// Address is an anyone-can-spend P2SH script
+	address, scriptSig, err := GenerateAnyoneCanSpendAddress(&chainParams)
+	if err != nil {
+		t.Fatalf("Error constructing P2SH address: %v", err)
+	}
+
+	genesisBlock := btcutil.NewBlock(chainParams.GenesisBlock)
+
+	// Generate block with spendable coinbase
+	blockVersion := int32(2)
+	timestamp := time.Now().Truncate(time.Second)
+	block, err := rpctest.CreateBlock(genesisBlock, nil, blockVersion,
+		timestamp, address, &chainParams)
+	if err != nil {
+		t.Fatalf("failed to generate block: %v", err)
+	}
+
+	// Process first block with spendable output
+	isOrphan, err := syncMgr.ProcessBlock(block, blockchain.BFNone)
+	if err != nil {
+		t.Fatalf("failed to process block: %v", err)
+	}
+	if isOrphan {
+		t.Fatal("failed to accept valid block")
+	}
+
+	// Create valid transaction
+	tx0, err := block.Tx(0)
+	if err != nil {
+		t.Fatal(err)
+	}
+	tx1, err := createSpendingTx(tx0, 0, scriptSig, address)
+	if err != nil {
+		t.Fatal(err)
+	}
+
+	// Attempt to process transaction
+	syncChan := make(chan struct{})
+	syncMgr.QueueTx(tx1, localNode, syncChan)
+	select {
+	case <-syncChan:
+	case <-time.After(time.Second):
+		t.Fatal("Timeout waiting for sync manager to process transaction 1")
+	}
+
+	// Assert transaction was accepted by checking that call was made to notify
+	// peers of the new transaction.
+	select {
+	case call := <-ctx.peerNotifier.announceNewTransactionsChan:
+		if len(call.newTxs) != 1 ||
+			!call.newTxs[0].Tx.Hash().IsEqual(tx1.Hash()) {
+
+			t.Fatalf("PeerNotifier received unexpected "+
+				"AnnounceNewTransactions call: %v", call.newTxs)
+		}
+	default:
+		t.Fatalf("Expected SyncManager to make " +
+			"AnnounceNewTransactions call to PeerNotifier")
+	}
+
+	// Create chain of two transactions and broadcast out of order
+	tx2, err := createSpendingTx(tx1, 0, scriptSig, address)
+	if err != nil {
+		t.Fatal(err)
+	}
+
+	tx3, err := createSpendingTx(tx2, 0, scriptSig, address)
+	if err != nil {
+		t.Fatal(err)
+	}
+
+	// Attempt to process orphan transaction
+	syncMgr.QueueTx(tx3, localNode, syncChan)
+	select {
+	case <-syncChan:
+	case <-time.After(time.Second):
+		t.Fatal("Timeout waiting for sync manager to process transaction 3")
+	}
+
+	// AnnounceNewTransactions should be called with no transactions
+	select {
+	case call := <-ctx.peerNotifier.announceNewTransactionsChan:
+		if len(call.newTxs) != 0 {
```

Comment:
This feels like a (rather benign) bug to me (not in your tests, rather in the code and your tests have found it which is awesome).  `AnnounceNewTransactions` is intended to announce transactions that are actually added to the mempool.  Orphan transactions are _not_ added to the mempool until/if their parent(s) show(s) up.  Also, there is little use to receiving and empty slice since the caller can't even tell what tx was added as an orphan if that were the intent (which it isn't as previously noted).

Since the code that is called in `server` and `rpcserver` both just do a for loop over the slice, which is empty, nothing actually happens, so the bug doesn't actually have any negative behavior.  Nevertheless, I feel like `handleTxMsg` should be doing the following to avoid sending the notification when no transactions were actually added:

```Go
if len(acceptedTxs) > 0 {
	sm.peerNotifier.AnnounceNewTransactions(acceptedTxs)
}
```

Naturally that means this test would need to be updated.